include/zephyr/storage/storage_area/storage_area.h

```diff
@@ -0,0 +1,227 @@
+
+#include <stdint.h>
+#include <stddef.h>
+#include <sys/types.h>
+#include <zephyr/sys/util.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Storage_area interface
+ * @defgroup storage_area_interface Storage_area interface
+ * @ingroup storage_apis
+ * @{
+ */
+
+/* Allow the sa_off_t type to be redefined if needed */
+typedef off_t sa_off_t;
+
+struct storage_area;
+
+struct storage_area_iovec { /* storage area io vector */
+	void *data;         /* pointer to data */
+	size_t len;         /* data length */
+};
+
+enum storage_area_properties_mask {
+	STORAGE_AREA_PROP_READONLY = BIT(0),
+	STORAGE_AREA_PROP_FOVRWRITE = BIT(1), /* full overwrite (ram, rram, ...) */
+	STORAGE_AREA_PROP_LOVRWRITE = BIT(2), /* limited overwrite (nor flash) */
+	STORAGE_AREA_PROP_ZEROERASE = BIT(3), /* erased value is 0x00 */
+	STORAGE_AREA_PROP_AUTOERASE = BIT(4), /* erase while writing */
+};
+
+enum storage_area_ioctl_cmd {
+	STORAGE_AREA_IOCTL_NONE,
+	STORAGE_AREA_IOCTL_XIPADDRESS, /* retrieve the storage area xip address */
+};
+
+/**
+ * @brief storage_area API
+ *
+ * API to access storage area.
+ */
+struct storage_area_api {
+	int (*readv)(const struct storage_area *area, sa_off_t offset,
+		     const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*writev)(const struct storage_area *area, sa_off_t offset,
+		    const struct storage_area_iovec *iovec, size_t iovcnt);
+	int (*erase)(const struct storage_area *area, size_t sblk, size_t bcnt);
+	int (*ioctl)(const struct storage_area *area,
+		     enum storage_area_ioctl_cmd cmd, void *data);
+};
+
+/**
+ * @brief storage_area struct
+ */
+struct storage_area {
+	const struct storage_area_api *api;
+	size_t write_size;
+	size_t erase_size;
+	size_t erase_blocks;
+	uint32_t props; /* bitfield of storage area properties */
+};
+
+/**
+ * @brief storage_area macros
+ */
+#define STORAGE_AREA_HAS_PROPERTY(area, prop) ((area->props & prop) == prop)
+#define STORAGE_AREA_WRITESIZE(area)          area->write_size
+#define STORAGE_AREA_ERASESIZE(area)          area->erase_size
+#define STORAGE_AREA_SIZE(area)							\
+	(area->erase_size * area->erase_blocks)
+#define STORAGE_AREA_READONLY(area)                                             \
+	STORAGE_AREA_HAS_PROPERTY(area, STORAGE_AREA_PROP_READONLY)
+#define STORAGE_AREA_FOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, STORAGE_AREA_PROP_FOVRWRITE)
+#define STORAGE_AREA_LOVRWRITE(area)                                            \
+	STORAGE_AREA_HAS_PROPERTY(area, STORAGE_AREA_PROP_LOVRWRITE)
+#define STORAGE_AREA_ERASEVALUE(area)                                           \
+	STORAGE_AREA_HAS_PROPERTY(area, STORAGE_AREA_PROP_ZEROERASE) ?		\
+	0x00 : 0xff
+#define STORAGE_AREA_AUTOERASE(area)						\
+	STORAGE_AREA_HAS_PROPERTY(area, STORAGE_AREA_PROP_AUTOERASE)
+
+#define GET_STORAGE_AREA(_name)							\
+	(struct storage_area *)&_storage_area_##_name.area
```

Comment:
Mixing tabs and spaces, I'd suggest formatting with clang-format here.

Reply:
Hi @pdgendt, thanks for your comments. I'm doing quite a lot of changes at the moment and I try to use clang-format but this is giving me a hard time. It doesn't respect line length (I set it to 81 because this fits my vision best) on macro's and it reformats macro's in a strange way (especially when filling structs). It also does not want to use tabs in macro's. But I will do my best to avoid mixing tabs and spaces.